### PR TITLE
Implement DiffusionLM-inspired embedding space collapse prevention

### PR DESCRIPTION
**Pull request by LiamConnell**
## Summary

Implements comprehensive embedding space collapse prevention based on DiffusionLM research, addressing the repetitive "from"/"no" token generation identified in Issue #20.

## 🔬 Problem Analysis

Our latest experiment revealed **embedding space collapse** - the model generates repetitive tokens because their embeddings become indistinguishable during training. This matches the DiffusionLM paper's finding: *"When directly adapting the diffusion objective to the embedding diffusion model, the embedding space will collapse."*

### **1. Anchor Loss for Embedding Regularization**
```python
def anchor_loss(predicted_embeddings, target_embeddings, alpha=0.1):
    """Prevents embedding space collapse by regularizing denoised embeddings."""
    return alpha * F.mse_loss(predicted_embeddings, target_embeddings)
```
- Creates stronger connection between denoised and ground truth embeddings
- Prevents token embeddings from becoming indistinguishable
- Configurable weight via `--anchor_weight` (default: 0.1)

### **2. Constant Rounding Weight Schedule**
```python
def constant_rounding_weight_schedule(epoch, total_epochs, weight=0.4):
    """Maintains consistent rounding loss to prevent embedding collapse."""
    return weight
```
- Replaces aggressive decay (0.5→0.05) with stable weight (0.4)
- Activated via `--constant_rounding_weight` flag
- Based on DiffusionLM's approach to maintain rounding loss throughout training

### **3. Embedding Space Health Monitoring**
```python
def embedding_anisotropy_score(embeddings):
    """Measures embedding diversity using cosine similarity variance."""
    # Higher scores = better embedding diversity (less collapse)
```
- Tracks embedding space health every 10 epochs
- Monitors token distinguishability during training
- Based on DiffusionLM's self-similarity anisotropy metric

## 📊 Training Configuration Updates

**Enhanced Training Args**:
```yaml
args: [
  "--rounding_weight", "0.4",           # Reduced from 0.5 for better balance
  "--anchor_weight", "0.1",             # NEW: Anchor loss weight
  "--constant_rounding_weight",         # NEW: Maintain stable rounding loss
  # ... existing args unchanged
]
```

## 🧪 Expected Outcomes

**Fixes Embedding Space Collapse**:
- ✅ Eliminates repetitive "from"/"no" token generation
- ✅ Maintains token embedding distinguishability throughout training
- ✅ Enables semantic clustering by syntactic role (POS tags)
- ✅ Produces diverse, coherent Shakespeare-style text

**Enhanced Monitoring**:
- Real-time anchor loss tracking alongside diffusion and rounding losses
- Embedding anisotropy scores for space health assessment
- Comprehensive loss logging for training analysis

## 🔬 Research Foundation

**Based on**: 
- "Diffusion-LM Improves Controllable Text Generation" (Li et al., 2022)
- "Empowering Diffusion Models on the Embedding Space for Text Generation"

**Key Insight**: *"Recent works of diffusion on embeddings introduce the rounding loss to address the challenge that when directly adapting the diffusion objective to the embedding diffusion model, the embedding space will collapse."*

## 🔗 Related Issues

**Addresses**: Issue #21 - Implement DiffusionLM-inspired embedding space collapse prevention  
**Builds On**: Issue #20 - Mode collapse prevention experiment (revealed embedding space issue)

## ✅ Testing Plan

Ready for immediate deployment:
1. **100-epoch experiment** with new embedding space protection
2. **Monitor embedding anisotropy** throughout training
3. **Validate text quality** - expect diverse, non-repetitive generation
4. **Compare against baseline** - should eliminate "from"/"no" dominance

---

This implementation directly addresses the root cause of our text generation quality issues and should enable high-quality, diverse Shakespeare-style output.

### CODE REVIEW

**Comment by LiamConnell** [comment]
## Implementation Updates: Memory Optimization & Bug Fixes

### Latest Changes Applied

#### 🔧 **Anchor Loss Scaling Fix**
- **Issue**: Anchor loss values were extremely high (500+ range) causing numerical instability
- **Fix**: Reduced alpha parameter from 0.1 → 0.01 (10x reduction)
- **Result**: Anchor loss now in reasonable range (0.01-68.6) during training

#### 💾 **Memory Optimization for Large Vocabularies** 
- **Issue**: `embedding_anisotropy_score()` caused CUDA OOM with 50K+ vocab size
- **Problem**: Computing full similarity matrix required ~244GB memory on V100
- **Fix**: Added random sampling to limit computation:
```python
def embedding_anisotropy_score(embeddings, sample_size=1000):
    # Sample embeddings to avoid memory issues with large vocabularies
    vocab_size = embeddings.size(0)
    if vocab_size > sample_size:
        indices = torch.randperm(vocab_size, device=embeddings.device)[:sample_size]
        embeddings = embeddings[indices]
```
- **Result**: Memory usage reduced from 244GB → <1GB while maintaining monitoring capability

#### ⚡ **Deployment Optimizations**
- **Batch Size**: Reduced from 8 → 4 for V100 memory efficiency
- **Container**: Fresh build with all fixes applied
- **Validation**: Local testing confirmed stable training

### Experiment Status

**Current Job**: `6714632384576225280` 🏃 **RUNNING**  
**Configuration**: DiffusionLM implementation with memory optimizations  
**Expected**: Stable 100-epoch training with embedding space monitoring

**Previous Failures**:
1. **Job `2237491405016530944`**: Anchor loss scaling issue → Fixed
2. **Job `1474131268177231872`**: Memory OOM during anisotropy calculation → Fixed

### Key Learnings

1. **Anchor Loss Sensitivity**: Small alpha values (0.01) work better than larger ones (0.1)
2. **Memory Management**: Large vocabulary similarity matrices require careful optimization
3. **Iterative Debugging**: Cloud deployment benefits from local testing validation

The implementation is now robust and ready for successful completion of the 100-epoch experiment.

**Monitor Progress**: `uv run python deployment/monitor.py 6714632384576225280 --logs`
